data/tokenizer.py

```diff
@@ -237,3 +237,103 @@ def untokenize_tgt(self, tokens: list[str]) -> list[int]:
         starts = [int(token[6:]) for token in tokens if token not in self.specials]
 
         return starts
+
+
+class MaskedMidiEncoder(MultiTokEncoder):
+    def __init__(self, quantization_cfg: DictConfig, keys: list[str] = None, masking_probability: float = 0.1):
+        super().__init__()
+        self.quantization_cfg = quantization_cfg
+        self.masking_probability = masking_probability
+        if keys is None:
+            self.keys = ["pitch", "start_bin", "duration_bin", "velocity_bin"]
+        else:
+            self.keys = keys
+
+        self.specials = ["<PAD>", "<MASK>"]
+
+        self.vocab = list(self.specials)
+
+        # add midi tokens to vocab
+        self._build_vocab()
+        self.token_to_id = {token: it for it, token in enumerate(self.vocab)}
+
+    def __rich_repr__(self):
+        yield "MaskedMidiEncoder"
+        yield "vocab_size", self.vocab_size
+
+    @property
+    def vocab_size(self) -> int:
+        return len(self.vocab)
+
+    def _build_vocab(self):
+        time_tokens_product = itertools.product(
+            # weird because we want to use dstart if there is a dstart_bin key, and start if there is a start_bin
+            range(self.quantization_cfg[self.keys[1][:-4]]),
+            range(self.quantization_cfg.duration),
+        )
+
+        for pitch in range(21, 109):
+            self.vocab.append(f"PITCH-{pitch}")
+        for start, duration in time_tokens_product:
+            self.vocab.append(f"TIME-{start:0.0f}-{duration:0.0f}")
```

Comment:
Does it make sense to try this wit separate `TIME` and `DURATION` tokens? 🤔

Reply:
So far I think it won't matter that much so we might as well seperate them.